app/src/org/commcare/dalvik/activities/EntitySelectActivity.java

```diff
@@ -221,25 +232,55 @@ public void onClick(View v) {
         searchbox.setHorizontallyScrolling(false);
         searchResultStatus = (TextView) findViewById(R.id.no_search_results);
         header = (LinearLayout)findViewById(R.id.entity_select_header);
-        
-        barcodeButton = (ImageButton)findViewById(R.id.barcodeButton);
 
         mViewMode = session.isViewCommand(session.getCommand());
 
-        barcodeButton.setOnClickListener(new OnClickListener() {
+        Callout callout = shortSelect.getCallout();
 
-            public void onClick(View v) {
-                Intent i = new Intent("com.google.zxing.client.android.SCAN");
-                try {
-                    startActivityForResult(i, BARCODE_FETCH);
-                } catch(ActivityNotFoundException anfe) {
-                    Toast noReader = Toast.makeText(EntitySelectActivity.this, "No barcode reader available! You can install one from the android market.", Toast.LENGTH_LONG);
-                    noReader.show();
+        if (callout == null) {
+            // Default to barcode scanning if no callout defined in the detail
+            calloutButton = (ImageButton)findViewById(R.id.barcodeButton);
+            calloutButton.setOnClickListener(new OnClickListener() {
+                public void onClick(View v) {
+                    Intent i = new Intent("com.google.zxing.client.android.SCAN");
+                    try {
+                        startActivityForResult(i, BARCODE_FETCH);
+                    } catch (ActivityNotFoundException anfe) {
+                        Toast noReader = Toast.makeText(EntitySelectActivity.this,
+                                "No barcode reader available! You can install one " +
+                                "from the android market.",
+                                Toast.LENGTH_LONG);
+                        noReader.show();
+                    }
                 }
+            });
+        } else {
+            CalloutData calloutData = callout.evaluate();

```

Comment:
I had to do some wonky merging (I don't know why) and just want to make sure Callout/CalloutData stuff here is correct.